rfcs/20210624-persistent-cache/README.md

```diff
@@ -0,0 +1,259 @@
+way.
+
+#### Interprocess and Thread Synchronization
+
+The access to the persistent cache should be performed in a process and thread
+safe manner. For performance reasons both process and thread synchronization
+should be performed using readers-writer lock.
+
+### Option 2: Provide API for Implementing Persistent Cache
+
+oneDNN provides an API so that users can implement persistent cache in their
+applications.
+
+#### API
+
+The API will be responsible for providing a user with an ID for the binary
+representation of a kernel and the binary representation itself.
+
+* ID is an array of bytes represented as `std::vector<unsigned char>` that is a
+unique identifier for a binary representation. ID should take into account that
+binaries compiled for different devices are different binaries. The binaries
+compiled for a sub-device considered compatible with the main device.
+* Binary representation is an array of bytes represented as `std::vector<unsigned char>`.
+Since oneDNN primitive can consist of multiple kernels the API return
+`std::vector<std::vector<unsigned char>>`.
+
+This proposal covers only providing binary representations of kernels. However,
+it's possible that oneDNN will support serialization of the primitives in the future
+so that users can cache the whole model on disk. Given that, it makes sense to
+make API more common and scalable.
+
+The suggestion is to introduce a binary kind to distinguish binary representation of
+a kernel from primitive.
+
+```cpp
+struct primitive {
+    enum class binary_kind {
+        kernel,
+        /*primitive - unimplemented */
+    };
+    /* ... */
+};
+```
+
+The API for obtaining ID will be provided by the `primitive_desc_base` class
+therefore the API will be available in each `primitive_desc` class.
+
+```cpp
+struct primitive_desc_base {
+    /* ... */
+    std::vector<unsigned char> get_id() const;
+};
+```
+
+The `primitive` class will get a new constructor to take a binary and new API
+to provide user with the binary.
+
+```cpp
+struct primitive {
+    /* ... */
+    primitive(const primitive_desc &pd, const std::vector<std::vector<unsigned char>> &binaries, binary_kind kind);
+    // Primitives that are not supported return an empty vector.
+    std::vector<std::vector<unsigned char>> get_binaries(binary_kind kind) const;
+};
+```
+
+##### API usage example
+
+```cpp
+using namespace dnnl;
+
+{
+    convolution_forward::primitive_desc conv_pd(desc, attr, engine);
+    primitive conv(conv_pd);
+
+    auto key = conv_pd.get_id();
+    auto binaries = conv.get_binaries(primitive::binary_kind::kernel);
+    // Store the binaries on disk.
+    store_to_persistent_cache(key, binaries);
+}
+
+{
+    convolution_forward::primitive_desc conv_pd(desc, attr, engine);
+    auto key = conv_pd.get_id();
+    // Load binaries from disk.
+    auto binaries = load_from_persistent_cache(key);
+    primitive conv_from_cache(pd, binary, primitive::binary_kind::kernel);
```

Comment:
Do you know how much time takes resource initialization like scales, zero points, etc. (map, fill, unmap) comparing to the kernel build?

Reply:
In the in-memory primitive cache the resource initialization happens every time when primitives fetched from the primitive cache. From what I remember the time of primitive creation in the case of cache hit is usually between 20-200ms. The average is about 50ms and the resource creation takes most of the time (like 95%).

So the primitive creation time using `primitive conv_from_cache(pd, binary, primitive::binary_kind::kernel);` will be the same as primitive creation in the case of cache hit (when binary is cached) if I'm not missing anything.